xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml

```diff
@@ -976,28 +977,193 @@ XWiki.Annotation = Class.create({
+   * Useful when the changes on annotations are done from the tab, when the document position should still stay in the
+   * tab
+   * @param force boolean specifying whether loading should be done even if there are no annotations to display (useful
+   * for deleting annotations, which should be reflected in the annotated element even if no annotations are still left
+   * to display)
+  */
+  loadAnnotations: function(annotations, andShow, navigateToPane, force) {
+    if (annotations.size() == 0 &amp;&amp; !force) {
+      return;
     }
+    // For avoiding adding the same annotation twice, all the annotations signatures are removed before adding them again.
+    this.removeAnnotationsAndSelectionSignatures();
+    var annSystem = this;
+    annotations.each(function(ann) {
+      var plainTextStartOffset = ann.fields.find(field =&gt; field.name == 'plainTextStartOffset').value
+      var plainTextEndOffset = ann.fields.find(field =&gt; field.name == 'plainTextEndOffset').value
+      var annDOMRange = annSystem.getDOMRange(annSystem.annotatedElement, plainTextStartOffset, plainTextEndOffset);
+      // Since the annotation could start at a specific offset, the node is split for not wrapping the whole text and a
+      // new range is created to recalculate the new ends.
+      var strictRange = annSystem.fixRangeEnds(annDOMRange);
+
+      // All the nodes inside this range are extracted. Each node it is wrapped inside the annotation marker span.
+      var nodes = annSystem.getTextNodesInRange(strictRange);
+      for(var i = 0; i &lt; nodes.length; i++) {
+        annSystem.markAnnotation(nodes[i], ann);
+      }
+    });
+    this.addAnnotationMarkers(annotations);
+    // Notify the content change.
+    document.fire('xwiki:dom:updated', {elements: [this.annotatedElement]});
+    // Also handle the tab 'downstairs' when the annotations list changes.
+    this.reloadTab(navigateToPane);
     if (andShow) {
       this.toggleAnnotations(true);
     }
   },
 
+  /**
+   * Surround this node with a span corresponding to it's annotation.
+   *
+   * @param markedNode the node that corresponds to the current annotation
+   * @param ann object holding information about the annotation
+  */
+  markAnnotation: function(markedNode, ann) {
+    var wrapper = document.createElement('span');
+    wrapper.addClassName('annotation');
+    wrapper.addClassName('ID' + ann.annotationId);
+
+    var parentNode = markedNode.parentElement;
+    parentNode.replaceChild(wrapper, markedNode);
+    wrapper.appendChild(markedNode);
+  },
+
+  /**
+   * For the first and last node, split the nodes at the known offset to not annotate the whole text. Create a new range
+   * with these new nodes.
+   *
+   * @param range the DOM range of the annotated text
+  */
+  fixRangeEnds: function(range) {
+    var strictRange = new Range();
+
+    var newStartNode = range.startContainer.splitText(range.startOffset);
```

Comment:
Would be nice to check if the offset is within the text node (not before the first or after the last character), otherwise the split will create useless empty text nodes.